gnd/src/main/java/com/google/android/gnd/ui/map/gms/BasemapSelectorMapAdapter.java

```diff
@@ -0,0 +1,212 @@
+  private final BehaviorSubject<Point> cameraPositionSubject = BehaviorSubject.create();
+  private final PublishSubject<Extent> extentsSubject = PublishSubject.create();
+
+  @Nullable private LatLng cameraTargetBeforeDrag;
+  private GeoJsonLayer extentSelectionLayer;
+  private HashMap<String, Extent> availableExtents;
+
+  public BasemapSelectorMapAdapter(GoogleMap map, Context context) {
+    this.map = map;
+    this.context = context;
+    map.setMapType(GoogleMap.MAP_TYPE_HYBRID);
+    UiSettings uiSettings = map.getUiSettings();
+    uiSettings.setRotateGesturesEnabled(false);
+    uiSettings.setTiltGesturesEnabled(false);
+    uiSettings.setMyLocationButtonEnabled(false);
+    uiSettings.setMapToolbarEnabled(false);
+    uiSettings.setCompassEnabled(false);
+    uiSettings.setIndoorLevelPickerEnabled(false);
+    map.setOnCameraIdleListener(this::onCameraIdle);
+    map.setOnCameraMoveStartedListener(this::onCameraMoveStarted);
+    map.setOnCameraMoveListener(this::onCameraMove);
+    onCameraMove();
+  }
+
+  @Override
+  public void renderExtentSelectionLayer() {
+    File file = new File(context.getFilesDir(), GEO_JSON_FILE);
+
+    try {
+      InputStream is = new FileInputStream(file);
+      BufferedReader buf = new BufferedReader(new InputStreamReader(is));
+      String line = buf.readLine();
+      StringBuilder sb = new StringBuilder();
+      while (line != null) {
+        sb.append(line).append("\n");
+        line = buf.readLine();
+      }
+
+      JSONObject geoJson = new JSONObject(sb.toString());
+      this.extentSelectionLayer = new GeoJsonLayer(map, geoJson);
+      extentSelectionLayer.addLayerToMap();
+      extentSelectionLayer.setOnFeatureClickListener(this::onExtentSelection);
+
+      Iterable<GeoJsonFeature> jsonFeatures = this.extentSelectionLayer.getFeatures();
+
+      for (GeoJsonFeature feature : jsonFeatures) {
+        availableExtents.put(
+            feature.getId(),
+            Extent.newBuilder().setId(feature.getId()).setState(Extent.State.NONE).build());
+      }
+
+      Log.d(TAG, "Extent selection layer successfully loaded");
+
+    } catch (IOException | JSONException e) {
+      Log.e(TAG, "Unable to load extent selection layer", e);
+    }
+  }
+
+  @Override
+  public void updateExtentSelections(ImmutableSet<Extent> extents) {
+    stream(extents.asList())
+        .filter(extent -> availableExtents.containsKey(extent))
+        .forEach(this::updateExtentSelectionState);
+  }
+
+  private void updateExtentSelectionState(Extent extent) {
+    availableExtents.put(extent.getId(), extent);
+    extentsSubject.onNext(extent);
+  }
+
+  private void onExtentSelection(Feature feature) {
+    GeoJsonFeature geoJsonFeature = (GeoJsonFeature) feature;
+    Extent extent = availableExtents.get(geoJsonFeature.getId());
+
+    switch (extent.getState()) {
+      case DOWNLOADED:
+        updateExtentSelectionState(
+            extent.toBuilder().setState(Extent.State.PENDING_REMOVAL).build());
```

Comment:
Please add TODO to refactor repeated code here.

Reply:
Done, thanks!